<a href="https://colab.research.google.com/github/robertmadril15/CS4410/blob/main/Homework7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#MNIST model

from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D
from tensorflow.keras.utils import to_categorical
import time

(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))

y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

cnn = Sequential()

cnn.add(Conv2D(filters=64,
               kernel_size=(3, 3),
               activation='relu',
               input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Conv2D(filters=128,
               kernel_size=(3, 3),
               activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Flatten())
cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))

cnn.compile(optimizer='adam',
            loss='categorical_crossentropy',
            metrics=['accuracy'])

start_time = time.time()

history = cnn.fit(
    X_train,
    y_train,
    epochs=5,
    batch_size=64,
    validation_split=0.1
)

training_time = time.time() - start_time

print(f"training time: {training_time:.2f} seconds") #show training time for MNIST

loss, accuracy = cnn.evaluate(X_test, y_test, verbose=0)

print(f"accuracy: {accuracy:.4f}") #show accuracy for MNIST
print(f"test loss: {loss:.4f}") #show loss for MNIST

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 121s 141ms/step - accuracy: 0.8893 - loss: 0.3490 - val_accuracy: 0.9840 - val_loss: 0.0577
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 117s 139ms/step - accuracy: 0.9851 - loss: 0.0453 - val_accuracy: 0.9888 - val_loss: 0.0435
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 115s 136ms/step - accuracy: 0.9905 - loss: 0.0293 - val_accuracy: 0.9882 - val_loss: 0.0417
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 111s 132ms/step - accuracy: 0.9937 - loss: 0.0196 - val_accuracy: 0.9905 - val_loss: 0.0384
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 114s 135ms/step - accuracy: 0.9949 - loss: 0.0158 - val_accuracy: 0.9913 - val_loss: 0.0355
training time: 578.52 seconds
accuracy: 0.9910
test loss: 0.0262


In [2]:
#fashion MNIST model

from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D
from tensorflow.keras.utils import to_categorical
import time

(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))

y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

cnn = Sequential()

cnn.add(Conv2D(filters=64,
               kernel_size=(3, 3),
               activation='relu',
               input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Conv2D(filters=128,
               kernel_size=(3, 3),
               activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Flatten())
cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))

cnn.compile(optimizer='adam',
            loss='categorical_crossentropy',
            metrics=['accuracy'])

start_time = time.time()

history = cnn.fit(
    X_train,
    y_train,
    epochs=5,
    batch_size=64,
    validation_split=0.1
)

training_time = time.time() - start_time

print(f"fashion training time: {training_time:.2f} seconds") #show training time for fashion MNIST

loss, accuracy = cnn.evaluate(X_test, y_test, verbose=0)

print(f"fashion accuracy: {accuracy:.4f}") #show accuracy for fashion MNIST
print(f"fashion test loss: {loss:.4f}") #show test loss for fashion MNIST

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 117s 137ms/step - accuracy: 0.7680 - loss: 0.6475 - val_accuracy: 0.8842 - val_loss: 0.3277
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 114s 134ms/step - accuracy: 0.8845 - loss: 0.3148 - val_accuracy: 0.8857 - val_loss: 0.2996
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 113s 134ms/step - accuracy: 0.9065 - loss: 0.2564 - val_accuracy: 0.9067 - val_loss: 0.2526
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 143s 136ms/step - accuracy: 0.9186 - loss: 0.2200 - val_accuracy: 0.9127 - val_loss: 0.2378
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 145s 139ms/step - accuracy: 0.9299 - loss: 0.1867 - val_accuracy: 0.9138 - val_loss: 0.2345
fashion training time: 656.93 seconds
fashion accuracy: 0.9072
fashion test loss: 0.2546


In [4]:
#MNIST model with added layer

from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D
from tensorflow.keras.utils import to_categorical
import time

(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))

y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

cnn = Sequential()

cnn.add(Conv2D(filters=64,
               kernel_size=(3, 3),
               activation='relu',
               input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Conv2D(filters=128,
               kernel_size=(3, 3),
               activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Flatten())
cnn.add(Dense(units=4096, activation='relu')) #added layer
cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))

cnn.compile(optimizer='adam',
            loss='categorical_crossentropy',
            metrics=['accuracy'])

start_time = time.time()

history = cnn.fit(
    X_train,
    y_train,
    epochs=5,
    batch_size=64,
    validation_split=0.1
)

training_time = time.time() - start_time

print(f"training time: {training_time:.2f} seconds") #show training time for MNIST with added layer

loss, accuracy = cnn.evaluate(X_test, y_test, verbose=0)

print(f"accuracy: {accuracy:.4f}") #show accuracy for MNIST with added layer
print(f"test loss: {loss:.4f}") #show loss for MNIST with added layer

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 382s 450ms/step - accuracy: 0.9213 - loss: 0.2493 - val_accuracy: 0.9847 - val_loss: 0.0522
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 394s 466ms/step - accuracy: 0.9871 - loss: 0.0390 - val_accuracy: 0.9883 - val_loss: 0.0389
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 428s 449ms/step - accuracy: 0.9928 - loss: 0.0228 - val_accuracy: 0.9910 - val_loss: 0.0379
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 389s 458ms/step - accuracy: 0.9944 - loss: 0.0184 - val_accuracy: 0.9905 - val_loss: 0.0347
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 386s 458ms/step - accuracy: 0.9954 - loss: 0.0147 - val_accuracy: 0.9898 - val_loss: 0.0417
training time: 2034.67 seconds
accuracy: 0.9891
test loss: 0.0418


16.1) The first cell is the convnet with the MNIST data, with a training time of ~578 seconds, 99% accuracy, and 0.02 test loss. The second cell is the convnet with the fashion data, with a training time of ~657 seconds, accuracy of 90%, and test loss of 0.25 The training time difference between the two datasets is over a full minute. The accuracy for the fashion dataset is lower becuase the complex images are harder to classify.

16.4) The third cell is the convnet with the MNIST data with an added dense layer with 4096 neurons. This model has a much larger training time of ~2034 seconds and a similar accuracy of ~99% with a slightly higher test loss. With the added layer, the training time increases significantly, and the accuracy does not improve at all showing that its addition costs high computation without accuracy benefit.